File: java/src/main/java/io/mendirl/aventofcode/java2020/Day1.java
package io.mendirl.aventofcode.java2020;

import com.google.common.collect.Sets;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class Day1 {

    static Integer calcul_step1(List<String> inputs) {
        var collect = inputs.stream().map(Integer::parseInt).collect(Collectors.toSet());
        return calcul(collect, 2);
    }

    static Integer calcul_step2(List<String> inputs) {
        var collect = inputs.stream().map(Integer::parseInt).collect(Collectors.toSet());
        return calcul(collect, 3);
    }

    static Integer calcul(Set<Integer> entries, int size) {
        return Sets.combinations(entries, size).stream()
                .filter(integers -> integers.stream().reduce(0, Integer::sum) == 2020)
                .findFirst()
                .map(integers -> integers.stream().reduce(1, (x, y) -> x * y))
                .orElse(0);
    }

}

File: kotlin/src/main/kotlin/io/mendirl/adventofcode/Day1.kt
package io.mendirl.adventofcode

import com.google.common.collect.Sets

class Day1 : Day {

    override fun step1(inputs: List<String>) = calcul(HashSet(inputs.toInt()), 2)

    override fun step2(inputs: List<String>) = calcul(HashSet(inputs.toInt()), 3)

    private fun calcul(inputs: Set<Int>, size: Int): Int {
        return inputs.combinations(size)
            .filter { it.sum() == 2020 }
            .map { it.reduce { x, y -> x * y } }
            .first()
    }
}

fun <T> Set<T>.combinations(size: Int): Set<Set<T>> {
    return Sets.combinations(this, size)
}
